File: next/src/client.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { PlatformClient } from "./client.js";
import type { X402Config } from "./config.js";
import { sign } from "./sign.js";

const CFG: X402Config = {
  keyId: "x402_test_k",
  secret: "x402sk_test_s",
  baseUrl: "https://plat.test",
  env: "sandbox",
};

test("post signs over the exact body bytes and the right path", async () => {
  let captured: { url: string; body: string; headers: Record<string, string> } | undefined;
  const fakeFetch = (async (url: string, init: RequestInit) => {
    captured = {
      url,
      body: init.body as string,
      headers: Object.fromEntries(new Headers(init.headers).entries()),
    };
    return new Response(JSON.stringify({ paymentRequired: true }), { status: 200 });
  }) as unknown as typeof fetch;

  const client = new PlatformClient(CFG, fakeFetch);
  const r = await client.challenge({ route: "/premium", method: "GET" });

  assert.equal(r.status, 200);
  assert.equal((r.body as { paymentRequired: boolean }).paymentRequired, true);
  assert.equal(captured!.url, "https://plat.test/api/v1/challenge");
  assert.equal(captured!.body, '{"route":"/premium","method":"GET"}');
  assert.equal(captured!.headers["content-type"], "application/json");
  assert.equal(captured!.headers["x-x402-key"], "x402_test_k");
  const ts = Number(captured!.headers["x-x402-timestamp"]);
  assert.equal(
    captured!.headers["x-x402-signature"],
    await sign("x402sk_test_s", "POST", "/api/v1/challenge", ts, captured!.headers["x-x402-nonce"], captured!.body),
  );
});

test("transport failure is status 0 (fail closed)", async () => {
  const fakeFetch = (async () => {
    throw new Error("boom");
  }) as unknown as typeof fetch;
  const r = await new PlatformClient(CFG, fakeFetch).verify({ route: "/p", method: "GET", nonce: "n" });
  assert.equal(r.status, 0);
  assert.deepEqual(r.body, {});
});

test("non-json response body becomes {}", async () => {
  const fakeFetch = (async () =>
    new Response("upstream exploded", { status: 500 })) as unknown as typeof fetch;
  const r = await new PlatformClient(CFG, fakeFetch).challenge({ route: "/p", method: "GET" });
  assert.equal(r.status, 500);
  assert.deepEqual(r.body, {});
});

test("json array/scalar/null response body is coerced to {}", async () => {
  for (const payload of ["[1,2,3]", "42", "null"]) {
    const fakeFetch = (async () =>
      new Response(payload, {
        status: 200,
        headers: { "content-type": "application/json" },
      })) as unknown as typeof fetch;
    const r = await new PlatformClient(CFG, fakeFetch).verify({
      route: "/p",
      method: "GET",
      nonce: "n",
    });
    assert.equal(r.status, 200);
    assert.deepEqual(r.body, {});
  }
});

File: go/client_test.go
package x402

import (
	"errors"
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (f roundTripFunc) RoundTrip(r *http.Request) (*http.Response, error) { return f(r) }

func newResp(status int, body string) *http.Response {
	return &http.Response{
		StatusCode: status,
		Body:       io.NopCloser(strings.NewReader(body)),
		Header:     make(http.Header),
	}
}

func testCfg() Config {
	return Config{KeyID: "x402_test_k", Secret: "x402sk_test_s", BaseURL: "https://plat.test", Env: "sandbox"}
}

func TestPostSignsExactOrderedBytesAndPath(t *testing.T) {
	var gotURL, gotBody string
	var gotHeaders http.Header
	rt := roundTripFunc(func(r *http.Request) (*http.Response, error) {
		gotURL = r.URL.String()
		b, _ := io.ReadAll(r.Body)
		gotBody = string(b)
		gotHeaders = r.Header
		return newResp(200, `{"paymentRequired":true}`), nil
	})
	c := NewPlatformClient(testCfg(), &http.Client{Transport: rt})
	res := c.Challenge(ChallengeReq{Route: "/premium", Method: "GET"})

	if res.Status != 200 || res.Body["paymentRequired"] != true {
		t.Fatalf("unexpected response: %+v", res)
	}
	if gotURL != "https://plat.test/api/v1/challenge" {
		t.Fatalf("url = %q", gotURL)
	}
	// Struct field order preserved, slash not escaped, no HTML-escape, no trailing newline.
	if gotBody != `{"route":"/premium","method":"GET"}` {
		t.Fatalf("body bytes = %q", gotBody)
	}
	if gotHeaders.Get("Content-Type") != "application/json" {
		t.Fatalf("content-type = %q", gotHeaders.Get("Content-Type"))
	}
	ts := gotHeaders.Get("X-X402-Timestamp")
	var tsi int64
	for _, ch := range ts {
		tsi = tsi*10 + int64(ch-'0')
	}
	if gotHeaders.Get("X-X402-Signature") != Sign("x402sk_test_s", "POST", "/api/v1/challenge", tsi, gotHeaders.Get("X-X402-Nonce"), gotBody) {
		t.Fatal("signature does not match exact sent bytes")
	}
}

func TestVerifyOmitsAbsentOptionalFieldsAndEscapesNothing(t *testing.T) {
	var gotBody string
	rt := roundTripFunc(func(r *http.Request) (*http.Response, error) {
		b, _ := io.ReadAll(r.Body)
		gotBody = string(b)
		return newResp(200, `{"allowed":true}`), nil
	})
	c := NewPlatformClient(testCfg(), &http.Client{Transport: rt})
	c.Verify(VerifyReq{Route: "/п", Method: "GET", Nonce: "n"})
	// payer + payment_proof omitted (omitempty); unicode raw UTF-8 (no \u escaping); & < > not HTML-escaped.
	if gotBody != `{"route":"/п","method":"GET","nonce":"n"}` {
		t.Fatalf("verify body = %q", gotBody)
	}
}

func TestTransportErrorIsStatusZeroFailClosed(t *testing.T) {
	rt := roundTripFunc(func(r *http.Request) (*http.Response, error) {
		return nil, errors.New("boom")
	})
	c := NewPlatformClient(testCfg(), &http.Client{Transport: rt})
	res := c.Verify(VerifyReq{Route: "/p", Method: "GET", Nonce: "n"})
	if res.Status != 0 || len(res.Body) != 0 {
		t.Fatalf("expected fail-closed {0,{}}, got %+v", res)
	}
}

func TestNonObjectJSONBecomesEmptyMap(t *testing.T) {
	for _, payload := range []string{`[1,2,3]`, `42`, `null`, `"hi"`, `not json`} {
		rt := roundTripFunc(func(r *http.Request) (*http.Response, error) {
			return newResp(500, payload), nil
		})
		c := NewPlatformClient(testCfg(), &http.Client{Transport: rt})
		res := c.Challenge(ChallengeReq{Route: "/p", Method: "GET"})
		if res.Status != 500 || len(res.Body) != 0 {
			t.Fatalf("payload %q: expected {500,{}}, got %+v", payload, res)
		}
	}
}
